Game should be stoppable

TheSecretSquad opened the pull request:
When should we consider test cases such as sending game.start() to an already started game and sending game.stop() to a stopped game?

jamesladd commented:
Does it make sense?
Do we care right now?

I think yes but is ‘now’ the right time ?

Sent from my Commodore 64

> On 23 Nov 2017, at 4:54 pm, Pete DiSalvo <notifications@github.com> wrote:
> 
> When should we consider test cases such as sending game.start() to an already started game and sending game.stop() to a stopped game?
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/jamesladd/GameKata/pull/5
> 
> Commit Summary
> 
> Make TicTacToe class package-private.
> Test shouldBeStoppable.
> File Changes
> 
> M src/main/java/com/thingo/game/Game.java (2)
> M src/main/java/com/thingo/game/GameListener.java (2)
> M src/main/java/com/thingo/game/impl/TicTacToe.java (8)
> M src/test/java/com/thingo/game/impl/TicTacToeTest.java (15)
> Patch Links:
> 
> https://github.com/jamesladd/GameKata/pull/5.patch
> https://github.com/jamesladd/GameKata/pull/5.diff
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

TheSecretSquad commented:
I guess we can continue without those tests for now.

jamesladd commented:
By sure I understand 

Sent from my Commodore 64

> On 25 Nov 2017, at 2:52 pm, Pete DiSalvo <notifications@github.com> wrote:
> 
> I guess we can continue without those tests for now.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

TheSecretSquad commented:
Would it make sense to generalize to something like this:

```java
private void ifStateDo(boolean state, Runnable action) {
    if (state) {
        action.run();
    }
}
```

Would be used like this:

```java
ifStateDo(isStarted(), () -> { throw new RuntimeException("Game is already started."); });
```

jamesladd commented:
I think that is one approach however that is hiding the issue behind a Lambda - I do like the thinking though.

Here is a hint - Proxy Pattern

Sent from my Commodore 64

> On 28 Nov 2017, at 4:31 pm, Pete DiSalvo <notifications@github.com> wrote:
> 
> Would it make sense to generalize to something like this:
> 
> private void ifStateDo(boolean state, Runnable action) {
>     if (state) {
>         action.run();
>     }
> }
> Would be used like this:
> 
> ifStateDo(isStarted(), () -> { throw new RuntimeException("Game is already started."); });
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

TheSecretSquad commented:
We could create another interface representing the state of the game. It could have methods like game's `start()` and `stop()`. Although each method would still have an if statement, so we're not removing the if statements, just moving them.

Another option is to create an interface representing the state of the game, but instead implement it to return new states when calling `start` and `stop`. The other states could be implemented using subclasses. That way the subclasses would handle the if statements polymorphically instead.

Not sure how the changes would be communicated the listener in either of these cases.

jamesladd commented:
That wouldn’t be a proxy ;)


Sent from my Commodore 64

> On 28 Nov 2017, at 5:00 pm, Pete DiSalvo <notifications@github.com> wrote:
> 
> We could create another interface representing the state of the game. It could have methods like game's start() and stop(). Although each method would still have an if statement, so we're not removing the if statements, just moving them.
> 
> Another option is to create an interface representing the state of the game, but instead implement it to return new states when calling start and stop. The other states could be implemented using subclasses. That way the subclasses would handle the if statements polymorphically instead.
> 
> Not sure how the changes would be communicated the listener in either of these cases.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

jamesladd commented:
How about you create an internal instance of Game that is the only field of TicTacToe.
This instance is delegated all calls.
When the TicTacToe is created this internal field w be an instance of UnstartedGame and when the Game is started this would be an instance of Started game - all the TicTacToe methods do is invoke the same method on this field and notify the listener

Eg

Start() {
  this.delegate = this.delegate.start();
  listener.started(this);
  return this;
}

This will remove all if statements 

Sent from my Commodore 64

> On 28 Nov 2017, at 5:00 pm, Pete DiSalvo <notifications@github.com> wrote:
> 
> We could create another interface representing the state of the game. It could have methods like game's start() and stop(). Although each method would still have an if statement, so we're not removing the if statements, just moving them.
> 
> Another option is to create an interface representing the state of the game, but instead implement it to return new states when calling start and stop. The other states could be implemented using subclasses. That way the subclasses would handle the if statements polymorphically instead.
> 
> Not sure how the changes would be communicated the listener in either of these cases.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

jamesladd commented:
Try this approach .... it models the state with Inner classes, which means we don't need 'if' statements around the state as polymorphism will take care of that, the state classes mean we can also be more immutable.

public class TicTacToe implements Game {

    private final GameListener gameListener;
    private Game state;

    TicTacToe(GameListener gameListener) {
        this.gameListener = gameListener;
        this.state = new TicTacToeInitialState();
    }

    @Override
    public Game start() {
        state = state.start();
        gameListener.started(this);
        return this;
    }

    // State classes model state and remove most 'if' statements
    // which are typically there to query state.
    //
    // Also helps with immutability to capture state in a new instance.

    private class TicTacToeInitialState implements Game {
        @Override
        public Game start() {
            return new TicTacToeStartedState();
        }
    }

    private class TicTacToeStartedState implements Game {
        @Override
        public Game start() {
            throw new RuntimeException("Game already started");
        }
    }
}

TheSecretSquad commented:
I like where you're going with this. I have some questions though.

**Question 1**
I'm familiar with the proxy pattern. I'm also familiar with the concept of replacing conditionals with polymorphism. I'm not sure why I didn't think of this first.

What was going through your mind that lead you to this solution? Did you just apply known patterns, like replace conditional with polymorphism? I don't like thinking about it as replacing if statements, because this means we're not thinking in objects. We have to think procedurally and then fix it. Do you ask how the problem can be solved by sending a message or using an object? Is it possible to take a concept like "when we send the start message, if the game is already started, an exception should be thrown" and map that to an object-oriented solution? Can we say that a condition requires a decision be made and objects make decisions, so we need an object to make the decision? I'm just throwing ideas out there.

**Question 2**
Is it confusing that the Game interface is used to represent both the concept of a whole game and also the state part of a game? Especially when considering that TicTacToe implements Game by returning itself, while the State objects can return new objects. This means when using a TicTacToe object through the game interface it's okay to say

game.start();
game.stop();
game.start();

With the State objects, we have to do

state = state.start();
state = state.stop();
state = state.start();

They can't be used interchangeably through the Game interface, which is why I would have thought to use the state pattern and create a separate interface called GameState or State that has immutable semantics.

jamesladd commented:
>> I don't like thinking about it as replacing if statements, because this means we're not thinking in objects. 

I think removing if statements is thinking in objects given the object can do the right thing w polymorphism rather than an if.

You could change the state Object inside TicTacToe to implement another interface since it is always internal to TicTacToe BUT the state = state.reasonToChange is important as this allows the state to be immutable - I’d avoid getters or setters on this state Object as much as possible.

The approach we are taking helps keep the separation of concerns - TicTacToe gives the behaviour and the internal state Object keeps the state - separate concerns.

Keep the questions coming 

Sent from my Commodore 64

> On 29 Nov 2017, at 2:16 pm, Pete DiSalvo <notifications@github.com> wrote:
> 
> I don't like thinking about it as replacing if statements, because this means we're not thinking in objects.

TheSecretSquad commented:
I implemented start and stop to toggle between `TicTacToeStartedState` and `TicTacToeInitialState` for now, but it seems like we might need other states there later (we may want a `TicTacToeStoppedState`, to indicate the game was started and then stopped, compared to initial state which means never started).

---

These changes are an example of something I encounter pretty often. We have the `GameListener` with `started()` and `stopped()` methods. We introduced the `State` objects, but they are an implementation detail that I'm not sure we want to reveal. However, adding a new `State` implementation likely means we will add a method to `GameListener`. Does it make sense to add methods to `GameListener` and keep `State` objects a hidden detail, or does it make sense to get rid of the specific methods from `GameListener` and just put one method on `GameListener`, e.g.
```Java
public interface GameListener {
    GameListener ChangedTo(State state);
}
```

jamesladd commented:
Don’t share an internal detail like state w a behavioural item like GameListener 

They are separate for a reason :)

You are asking good questions 

Sent from my Commodore 64

> On 1 Dec 2017, at 2:09 pm, Pete DiSalvo <notifications@github.com> wrote:
> 
> I implemented start and stop to toggle between TicTacToeStartedState and TicTacToeInitialState for now, but it seems like we might need other states there later (we may want a TicTacToeStoppedState, to indicate the game was started and then stopped, compared to initial state which means never started).
> 
> These changes are an example of something I encounter pretty often. We have the GameListener with started() and stopped() methods. We introduced the State objects, but they are an implementation detail that I'm not sure we want to reveal. However, adding a new State implementation likely means we will add a method to GameListener. Does it make sense to add methods to GameListener and keep State objects a hidden detail, or does it make sense to get rid of the specific methods from GameListener and just put one method on GameListener, e.g.
> 
> public interface GameListener {
>     GameListener ChangedTo(State state);
> }
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

TheSecretSquad commented:
I notice you've made a point of calling certain objects behavioral and the state as a "detail".
e.g.,
> ...TicTacToe gives the behaviour and the internal state Object keeps the state...

> ...internal detail like state w a behavioural item like GameListener

My understanding is all objects expose behavior and are "behavioral". Could you expand on why you make the distinction here? Is it any coincidence that the "behavioral" objects are mutable and the other is immutable? How do we then determine which things can/should be shared with other things and which shouldn't?

---

What should be our next step? Is there more to do on this, or should we merge it and continue?

jamesladd commented:
Good questions.

Behaviour is exposed to other behavioural objects but an object that encapsulates state does not or should not. It is confusing because they are both objects but have diff purposes. Thing Atom vs it’s nucleus 

At some point we simply have to change state or our app won’t do anything. The approach we are using means when state does change it is very controlled - I really like the way you have the Tictactoe state objects - I think it is really clean and clear when you say:  

state = new State()


I’ll think about what to do next ;)


Sent from my Commodore 64

> On 1 Dec 2017, at 3:06 pm, Pete DiSalvo <notifications@github.com> wrote:
> 
> I notice you've made a point of calling certain objects behavioral and the state as a "detail".
> e.g.,
> 
> ...TicTacToe gives the behaviour and the internal state Object keeps the state...
> 
> ...internal detail like state w a behavioural item like GameListener
> 
> My understanding is all objects expose behavior and are "behavioral". Could you expand on why you make the distinction here? Is it any coincidence that the "behavioral" objects are mutable and the other is immutable? How do we then determine which things can/should be shared with other things and which shouldn't?
> 
> What should be our next step? Is there more to do on this, or should we merge it and continue?
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>
